### PR TITLE
Potential fix for disabeling core modules

### PR DESCRIPTION
**Pull request by MatthijsKrempel**
You can add set the IsAlwaysRunning property to true in your FeatureAttribute or ModuleAttribute to make sure your feature can't be disable the feature.

This is a fix for https://github.com/OrchardCMS/OrchardCore/issues/3403

Feedback is welcome

### CODE REVIEW

**Comment by jptissot** [comment]
Looks good to me

CC @jtketch

**Comment by jtkech** [comment]
Need to leave soon, i will review it this night

But as i have already seen LGTM !

File: src/OrchardCore/OrchardCore.Abstractions/Modules/Manifest/ManifestAttributes.cs [hunk 1]
```diff
@@ -85,6 +85,11 @@ public FeatureAttribute()
         /// Set to <c>true</c> to only allow the Default tenant to enable it.
         /// </summary>
         public bool DefaultTenantOnly { get; set; }
+
+        /// <summary>
+        /// If the feature is always enabled, you can't disabled it
```
Comment (hishamco) [inline src/OrchardCore/OrchardCore.Abstractions/Modules/Manifest/ManifestAttributes.cs]: /// Check whether the feature is always enabled. Defaults to <c>false</c>.

File: test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs [hunk 2]
```diff
@@ -97,7 +98,7 @@ public TestThemeExtensionInfo(string name)
                 var features =
                     new List<IFeatureInfo>()
                     {
-                        {new FeatureInfo(name, name, 0, "", "", this, new string[0], false)}
+                        {new FeatureInfo(name, name, 0, "", "", this, new string[0], false, false)}
```
Comment (hishamco) [inline test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs]: Could we format this
```csharp
{  new FeatureInfo(name, name, 0, String.Empty, String.Empty, this, new string[0], false, false)  }
```

File: test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs [hunk 3]
```diff
@@ -124,7 +125,7 @@ public TestThemeExtensionInfo(string name, IFeatureInfo baseTheme)
                 Features =
                     new List<IFeatureInfo>()
                     {
-                        {new FeatureInfo(name, name, 0, "", "", this, new string[] { baseTheme.Id }, false)}
+                        {new FeatureInfo(name, name, 0, "", "", this, new string[] { baseTheme.Id }, false, false)}
```
Comment (hishamco) [inline test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs]: Same as above

File: test/OrchardCore.Tests/Shell/ShellContainerFactoryTests.cs [hunk 4]
```diff
@@ -145,7 +145,7 @@ private ShellBlueprint CreateBlueprint()
 
         public static IFeatureInfo AddStartup(ShellBlueprint shellBlueprint, Type startupType)
         {
-            var featureInfo = new FeatureInfo(startupType.Name, startupType.Name, 1, "Tests", null, null, null, false);
+            var featureInfo = new FeatureInfo(startupType.Name, startupType.Name, 1, "Tests", null, null, null, false, false);
```
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellContainerFactoryTests.cs]: Also here

Comment (MatthijsKrempel) [reply test/OrchardCore.Tests/Shell/ShellContainerFactoryTests.cs]: @hishamco I don't understand, what would like done?

File: test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs (hunk 3 above)
Comment (hishamco) [reply test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs]:
```csharp
{  new FeatureInfo(name, name, 0, "", "", this, new string[] { baseTheme.Id }, false, false)  }
```

File: test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs [hunk 5]
```diff
@@ -64,7 +64,7 @@ public TestModuleExtensionInfo(string name)
                 var features =
                     new List<IFeatureInfo>()
                     {
-                        {new FeatureInfo(name, name, 0, "", "", this, new string[0], false, false)}
+                        {new FeatureInfo(name, name, 0, String.Empty, String.Empty, this, new string[0], false, false)}
```
Comment (hishamco) [inline test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs]: Just format the object by add required spaces between the { }

File: test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs [hunk 6]
```diff
@@ -22,13 +22,23 @@
 
 namespace OrchardCore.Tests.DisplayManagement.Decriptors
 {
+    public static class StringArrayExtensions
```
Comment (hishamco) [inline test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs]: `StringArray` instead, coz these are not extension methods

File: test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs [hunk 7]
```diff
@@ -97,7 +108,7 @@ public TestThemeExtensionInfo(string name)
                 var features =
                     new List<IFeatureInfo>()
                     {
-                        {new FeatureInfo(name, name, 0, "", "", this, new string[0], false)}
+                        {new FeatureInfo(name, name, 0, "", "", this, StringArrayExtensions.Empty, false, false)}
```
Comment (hishamco) [inline test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs]: Missing space

File: test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs [hunk 8]
```diff
@@ -124,7 +135,7 @@ public TestThemeExtensionInfo(string name, IFeatureInfo baseTheme)
                 Features =
                     new List<IFeatureInfo>()
                     {
-                        {new FeatureInfo(name, name, 0, "", "", this, new string[] { baseTheme.Id }, false)}
+                        {new FeatureInfo(name, name, 0, String.Empty, String.Empty, this, new string[] { baseTheme.Id }, false, false)}
```
Comment (hishamco) [inline test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs]: Missing space

File: test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs [hunk 9]
```diff
@@ -22,20 +22,31 @@
 
 namespace OrchardCore.Tests.DisplayManagement.Decriptors
 {
+    public static class StringArray
+    {
+        private static readonly string[] _emptyStringArray = new string[0];
```
Comment (hishamco) [inline test/OrchardCore.Tests/DisplayManagement/Decriptors/DefaultShapeTableManagerTests.cs]: Use `_empty`

**Pull request merged**